engine/jobmaster/dm/ddl_coordinator.go

```diff
@@ -0,0 +1,841 @@
+			tables = c.tables[item.TargetTable]
+		}
+		tables[item.SourceTable] = struct{}{}
+		c.mu.Unlock()
+	case metadata.DropTable:
+		c.mu.Lock()
+		delete(c.tables[item.TargetTable], item.SourceTable)
+		if len(c.tables[item.TargetTable]) == 0 {
+			delete(c.tables, item.TargetTable)
+		}
+		c.mu.Unlock()
+	}
+	return ddls, conflictStage, err
+}
+
+// ShowDDLLocks show ddl locks.
+func (c *DDLCoordinator) ShowDDLLocks(ctx context.Context) ShowDDLLocksResponse {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	ddlLocks := make(map[metadata.TargetTable]DDLLock)
+	for targetTable, g := range c.shardGroups {
+		tbs := g.showTables()
+		ddlLocks[targetTable] = DDLLock{ShardTables: tbs}
+	}
+	return ShowDDLLocksResponse{Locks: ddlLocks}
+}
+
+func (c *DDLCoordinator) loadOrCreateShardGroup(ctx context.Context, targetTable metadata.TargetTable, jobCfg *config.JobCfg) (*shardGroup, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if g, ok := c.shardGroups[targetTable]; ok {
+		return g, nil
+	}
+
+	newGroup, err := newShardGroup(ctx, c.jobID, jobCfg, targetTable, c.tables[targetTable], c.kvClient, c.tableAgent)
+	if err != nil {
+		return nil, err
+	}
+	c.shardGroups[targetTable] = newGroup
+	return newGroup, nil
+}
+
+func (c *DDLCoordinator) removeShardGroup(ctx context.Context, targetTable metadata.TargetTable) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if g, ok := c.shardGroups[targetTable]; ok {
+		if err := g.clear(ctx); err != nil {
+			c.logger.Error("clear shard group failed", zap.Error(err))
+		}
+	}
+	delete(c.shardGroups, targetTable)
+}
+
+type shardGroup struct {
+	mu                  sync.RWMutex
+	normalTables        map[metadata.SourceTable]string
+	conflictTables      map[metadata.SourceTable]string
+	tableAgent          TableAgent
+	droppedColumnsStore *metadata.DroppedColumnsStore
+	id                  frameModel.MasterID
+	cfg                 *config.JobCfg
+	deleted             bool
+}
+
+func newShardGroup(ctx context.Context, id frameModel.MasterID, cfg *config.JobCfg, targetTable metadata.TargetTable, sourceTables map[metadata.SourceTable]struct{}, kvClient metaModel.KVClient, tableAgent TableAgent) (*shardGroup, error) {
+	g := &shardGroup{
+		tableAgent:          tableAgent,
+		normalTables:        make(map[metadata.SourceTable]string),
+		conflictTables:      make(map[metadata.SourceTable]string),
+		droppedColumnsStore: metadata.NewDroppedColumnsStore(kvClient, targetTable),
+		id:                  id,
+		cfg:                 cfg,
+	}
+	for sourceTable := range sourceTables {
+		stmt, err := g.tableAgent.FetchTableStmt(ctx, id, cfg, sourceTable)
+		// NOTE: There are cases where the source table exists but the table info does not,
+		// such as a table is created upstream when the coordinator starts, but the dm-worker is not yet synchronized
+		// we skip these tables in handleDDL.
+		// TODO: better error handling
+		if err != nil && !strings.HasPrefix(err.Error(), "table info not found") {
+			return nil, err
+		}
```

Comment:
should we continue the rest of tables?

Reply:
we just return the error. when new ddl comes, we will try to load tables again